crates/bevy_ecs/src/schedule/executor_parallel.rs

```diff
@@ -0,0 +1,508 @@
+                        // Wait until at least one system has finished.
+                        let index = self
+                            .finish_receiver
+                            .recv()
+                            .await
+                            .unwrap_or_else(|error| unreachable!(error));
+                        self.process_finished_system(index);
+                        // Gather other systems than may have finished.
+                        while let Ok(index) = self.finish_receiver.try_recv() {
+                            self.process_finished_system(index);
+                        }
+                        // At least one system has finished, so active access is outdated.
+                        self.rebuild_active_access();
+                    }
+                    self.update_counters_and_queue_systems();
+                }
+            });
+        });
+    }
+}
+
+impl ParallelExecutor {
+    /// Updates access and recondenses the archetype component bitsets of systems.
+    fn update_access(&mut self, systems: &mut [ParallelSystemContainer], world: &mut World) {
+        let mut all_archetype_components = HashSet::default();
+        for container in systems.iter_mut() {
+            let system = container.system_mut();
+            system.update_access(world);
+            if let Some(archetype_components) =
+                system.archetype_component_access().all_distinct_types()
+            {
+                all_archetype_components.extend(archetype_components);
+            }
+        }
+        let all_archetype_components = all_archetype_components.drain().collect::<Vec<_>>();
+        for (index, container) in systems.iter().enumerate() {
+            let system = container.system();
+            if !system.archetype_component_access().reads_all() {
+                self.system_metadata[index].archetype_component_access = system
+                    .archetype_component_access()
+                    .condense(&all_archetype_components);
+            }
+        }
+    }
+
+    /// Populates `should_run` bitset, spawns tasks for systems that should run this iteration,
+    /// queues systems with no dependencies to run (or skip) at next opportunity.
+    fn prepare_systems<'scope>(
+        &mut self,
+        scope: &mut Scope<'scope, ()>,
+        systems: &'scope [ParallelSystemContainer],
+        world: &'scope World,
+        resources: &'scope Resources,
+    ) {
+        self.should_run.clear();
+        for (index, system_data) in self.system_metadata.iter_mut().enumerate() {
+            // Spawn the system task.
+            if systems[index].should_run() {
+                self.should_run.set(index, true);
+                let start_receiver = system_data.start_receiver.clone();
+                let finish_sender = self.finish_sender.clone();
+                let system = unsafe { systems[index].system_mut_unsafe() };
+                let task = async move {
```

Comment:
As an experiment, I tried removing the start signal entirely by starting systems on demand from current scope after their criteria has been met. It resulted in significant performance wins on the low end:
* 99% for zero empty systems (but who cares)
* for every "empty system" variant < 50, 20%-0% increases (the wins got lower as the number of systems got heigher). After 50, it starts getting worse, with a max of ~20%
* I tried both crossbeam and async channels. They both had the same general characteristics, but crossbeam performed slightly better in most cases.
* non-send systems deadlock when using this approach.

Crossbeam + No Start Signal Results:
[crossbeam_removed_start.txt](https://github.com/bevyengine/bevy/files/5929663/crossbeam_removed_start.txt)

Reply:
I totally get why crossbeam would get worse with more systems (contention for atomic access). But its weird to me that `async_channel` also behaves the same way when swapped into the implementation above. (whereas it performs more favorably in the pr's current implementation). Probably because in the pr it awaits the channel receiver whereas with my changes it uses try_recv().

I'm not sure theres really any action to be taken here. Its better to optimize for many systems than for a few. Interestingly the `ecs_bench_suite` schedule benchmark appears to be completely unaffected by any of the approaches above, so in practice nobody will even notice the hit to smaller schedules.

I have a feeling that we could get a uniformly good perf boost if we removed the start channel while still awaiting the finished receiver, but I'm not sure the interfaces will allow for that.

Reply:
System tasks can be spawned as needed instead of beforehand, I didn't do that right now because it requires `Scope` to be `Send + Sync`, which it no longer is because it now lugs around `LocalExecutor`. I have a couple of rough ideas of how to handle that, but they would likely require messing with `bevy_tasks`, which I think is best done in a separate PR.

Non-send systems will likely have to keep using the start channels, because I don't think there's a lock-free way of spawning tasks on main thread from unspecified thread - which we have to account for because the coordination task is not guaranteed to always be on the main thread. We could spawn it locally, though; can't tell if there are perf implications without benchmarking it. Or I could be wrong and there is a simple lock-free way.

Reply:
Cool cool. Then I'm down to table the conversation for now. Perf is workable as-is, the api is what we want, and we have a number of "optimization levers" to pull.